Neos.ContentRepositoryRegistry/Classes/Factory/ProjectionCatchUpTrigger/SubprocessProjectionCatchUpTrigger.php

```diff
@@ -29,15 +38,80 @@ public function triggerCatchUp(Projections $projections): void
     {
         // modelled after https://github.com/neos/Neos.EventSourcing/blob/master/Classes/EventPublisher/JobQueueEventPublisher.php#L103
         // and https://github.com/Flowpack/jobqueue-common/blob/master/Classes/Queue/FakeQueue.php
-        foreach ($projections as $projection) {
-            Scripts::executeCommandAsync(
-                'neos.contentrepositoryregistry:subprocessprojectioncatchup:catchup',
-                $this->flowSettings,
-                [
-                    'contentRepositoryIdentifier' => $this->contentRepositoryId->value,
-                    'projectionClassName' => get_class($projection)
-                ]
-            );
+        $queuedProjections = array_map(fn($projection) => $this->startCatchUpWithQueueing($projection), iterator_to_array($projections->getIterator()));
+        $queuedProjections = array_filter($queuedProjections);
+
+        $attempts = 0;
+        while (!empty($queuedProjections)) {
+            usleep(random_int(100, 25000) + ($attempts * $attempts * 10)); // 50000μs = 50ms
+            if (++$attempts > 50) {
+                throw new \RuntimeException('TIMEOUT while waiting for projections to run queued catch up.', 1550232279);
+            }
+
+            foreach ($queuedProjections as $key => $projection) {
+                if ($this->catchUpStatesCache->has($this->cacheKeyPrefix($projection) . 'QUEUED') === false) {
```

Comment:
a rather minor detail, but I would suggest to remove the amount of magic keys by adding some
```php
private static function cacheEntryIdQueued(ProjectionInterface $projection): string {
  // ...
}
public static function cacheEntryIdRunning(ProjectionInterface $projection): string {
  // ...
}
```
the latter can be called from `SubprocessProjectionCatchUpCommandController::catchupCommand()`


Besides, it would make the logic more readable if we added dedicated methods for the cache interactions.
e.g. instead of
```php
if ($this->catchUpStatesCache->has($this->cacheKeyPrefix($projection) . 'RUNNING')) {
    return false;
}
$this->catchUpStatesCache->set($this->cacheKeyPrefix($projection) . 'RUNNING', 1);
```
something like
```php
if ($this->isCatchUpRunning($projection)) {
    return false;
}
$this->markCatchUpRunning($projection);
```

Reply:
Yeah, wasn't sure if static method in this place is _nice_ but I agree, I would be happier about that too.

Maybe I just make all of this a little class of it's own. I think that would be even nicer, `ProjectionCatchUpRunnerState($projection, $cache)` which then has stuff like ::isRunning() ::markRunning() ::markStopped()

Reply:
alright, update is on, looks much nicer now